expireable_store.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+type entry struct {
+	key   string
+	value *Quota
+
+	bucket int
+}
+
+type bucket struct {
+	entries map[string]*entry
+
+	expiresAt time.Time
+}
+
+// TODO document this, in particular provide some details around:
+// - the purpose and use of "buckets"
+type expirableStore struct {
+	maxSize int
+
+	items map[string]*entry
+
+	buckets            []bucket
+	bucketTTL          time.Duration
+	numberBuckets      int
+	nextBucketToExpire int
+
+	mu sync.Mutex
+
+	pool sync.Pool
+
+	cancelFunc context.CancelFunc
+	ctx        context.Context
+}
+
+func newExpireableStore(maxSize int, maxEntryTTL time.Duration, o ...Option) (*expirableStore, error) {
```

Comment:
```suggestion
func newExpirableStore(maxSize int, maxEntryTTL time.Duration, o ...Option) (*expirableStore, error) {
```
spelling

Reply:
I thought I had gotten all of these. They should all be fixed now, also fixed the filename.